packages/flutter/lib/src/widgets/tooltip.dart

```diff
@@ -0,0 +1,972 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+/// @docImport 'gesture_detector.dart';
+library;
+
+import 'dart:async';
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/gestures.dart';
+import 'package:flutter/rendering.dart';
+import 'package:flutter/services.dart';
+
+import 'basic.dart';
+import 'debug.dart';
+import 'feedback.dart';
+import 'framework.dart';
+import 'media_query.dart';
+import 'overlay.dart';
+import 'selection_container.dart';
+import 'ticker_provider.dart';
+
+const AnimationStyle _kDefaultAnimationStyle = AnimationStyle(
+  curve: Curves.fastOutSlowIn,
+  duration: Duration(milliseconds: 150),
+  reverseDuration: Duration(milliseconds: 75),
+);
+
+class _TooltipVisibilityScope extends InheritedWidget {
+  const _TooltipVisibilityScope({required super.child, required this.visible});
+
+  final bool visible;
+
+  @override
+  bool updateShouldNotify(_TooltipVisibilityScope old) {
+    return old.visible != visible;
+  }
+}
+
+/// Signature for building the tooltip overlay child.
+///
+/// The animation property exposes the underlying tooltip overlay child show and
+/// hide animation. This can be used to drive animations that sync up with the
+/// tooltip overlay child show/hide animation, for example to fade the tooltip
+/// in and out.
+typedef TooltipComponentBuilder =
+    Widget Function(BuildContext context, Animation<double> animation);
+
+/// Signature for computing the position of a tooltip.
+///
+/// The [TooltipPositionContext] contains all the necessary information for
+/// positioning the tooltip, including the target location, sizes, offset, and
+/// positioning preference.
+///
+/// Returns the offset from the top left of the overlay to the top left of the tooltip.
+///
+/// See also:
+///
+///  * [TooltipPositionContext], which contains the positioning parameters.
+typedef TooltipPositionDelegate = Offset Function(TooltipPositionContext context);
+
+/// Contextual information for positioning a tooltip.
+///
+/// This immutable data class contains all the necessary information for computing
+/// the position of a tooltip relative to its target widget.
+///
+/// See also:
+///
+///  * [TooltipPositionDelegate], which uses this context to compute tooltip positions.
+@immutable
+class TooltipPositionContext {
+  /// Creates a tooltip position context.
+  const TooltipPositionContext({
+    required this.target,
+    required this.targetSize,
+    required this.tooltipSize,
+    required this.verticalOffset,
+    required this.preferBelow,
```

Comment:
preferBelow is passed into the RawTooltip and piped here to be used by `TooltipPositionDelegate` seems unecessary, if material design tooltip has already known the preferBelow they could have uses it directly in `TooltipPositionDelegate` without going through the TooltipPositionContext